Checkout: Show taxes (Placeholder Edition)

deBhal opened the pull request:
This PR starts the process of adding checkout taxes.

We don't have real data available yet, and several important questions are still open, so the functionality is hidden behind the `show-tax` config flag. We're not ready to enable these even for the development config, so for now enable these using the flags query parameter, like `?flags=show-tax`


The transaction store is one of the remaining Flux stores ( see https://github.com/Automattic/wp-calypso/issues/24217 ). I'd love to find time to come back and reduxify it, but for I'm using `lib/redux-bridge` to forward the `FLUX_TRANSACTION_PAYMENT_SET` and  `TRANSACTION_NEW_CREDIT_CARD_DETAILS_SET` actions from Flux to Redux. The Payment country field was already wired up to a redux action, but this doesn't seem to cause any problems.

We're using a `httpData` request for the api call, and this effectively memoizes the request. For testing purposes, I've set a freshness of 2 days (to pluck a number out of the air). This means that you should only be able to make one successful request with a specific countryCode + postalCode combination. If you switch back to a previously used combination it won't fire off any extra requests, and once a value comes back you still won't get more requests (inside that 2 day period), you should just get the previously fetched value instantly.

Curiously, the `httpData` behaviour also seems to make it unnecessary to store the value in the redux state. I'm not quite sure what I think about that, but it's working, so...

We're calling out to the `/sites/example.wordpress.com/hello` endpoint until the tax-rate endpoint is ready in order to provide a http request to make the data-layer mechanisms run smoothly, and feeding in a dummy result if/when it returns successfully. The dummy result is the last three digits of the postcode after the decimal point (e.g. 12345 gives a 0.345 tax rate - 34.5%. An example real tax rate is 0.085).

**Testing Instructions**:

We're looking specifically for the `calypso_tax_rate_request` tracks event, along with a network request (to the placeholder address), but there are a bunch of things we want to check:

- Check that we do generate a tax event/network request (fetch) on 5 digit postcodes in the US
- Check that we never fetch with more or less than 5 digits, nor with anything other than numbers in there (it's fine if this includes valid postcodes for other countries, we're only concerned with the US for now)
- Check that we never fetch outside the US
- Check that we get both the country code and the postcode from stored credit cards.
- Check that we respond to the "New Credit/Debit Card" dialog as well as stored cards
- check that the `show-tax` config flag is required to see the tax rates, but not to trigger the http requests ( add url params: `?flags=show-tax`).

The other major concern is that we don't break anything else.

Two tricks to make this easier to verify, as shown in this gif:
1) `localStorage.setItem( 'debug', localStorage.debug + ',calypso:analytics:tracks' )`, and then optionally filter the log on `calypso_tax_rate_request`
2) pin `state.ui.payment` in redux tools so you can watch the values.
3) Change the `freshness` value in `requestTaxRate()` in `client/state/data-getters/index.js` to `60  * 1000` ( 1 minute) - long enough to see it work, but short enough to re-check against your stored credit cards if you want

Here, I've got 4 test credit-card pre-populated, 2 with different US postcodes, 1 from `AU` with a postcode that wouldn't match anyway, and one from `CA` with a 5 digit postcode:

![checkout httpdata](https://user-images.githubusercontent.com/5952255/47715274-f9d6c680-dc8a-11e8-86a3-305da2127ab6.gif)

Hywan commented:
Neat, thanks for starting that project!

prettyboymp commented:
I don't think we can hint one way or another about taxes on the pricing screen.  We won't be able to determine a tax rate or whether the price includes the tax or not until we have a billing zip code for the user.  

I think it's normal practice to just show the price of the item, whether it includes taxes or not.   For total prices, until we know the user's applicable tax rates,  we just need to label it as a `sub total`.

deBhal commented:
> I don't think we can hint one way or another about taxes on the pricing screen. We won't be able to determine a tax rate or whether the price includes the tax or not until we have a billing zip code for the user.

Yeah, I wanted to draw attention to the missing data so we can have this discussion.

For now (i.e. while we're focused on the US), we're going to drop all mention of tax from the plans screen, mostly out of design concerns.

Instead, we're going to concentrate/start with the checkout page where we've got the postcode, all the prices, and the "Pay $x" button.

deBhal commented:
I've updated the PR with data-layer middleware with tracks events, and wired up the payment details (using `redux-bridge`) and the length-picker, so I've been able to update the OP with a nice interactive demo now.

Don't look closely at the code yet, though ;)

a8ci18n commented on client/blocks/subscription-length-picker/index.jsx:
```diff
@@ -42,11 +45,34 @@ export class SubscriptionLengthPicker extends React.Component {
+		// taxRate unknown
+		if ( ! isNumber( taxRate ) ) {
+			// translators: This string is displayed immediately next to a localized price with a currency symbol, and is indicating that there may be an additional charge on top of the displayed price.
+			return translate( '+tax' );
```
Hi! I've found a possible matching string that has already been translated **54** times:
`translate( 'Tax' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented:
I've added the flux bridge for the stored cards as well, and refined how everything works, including putting the event outside the config flag and the display behind it.

There's still a niggling problem with the postal code field on the new card form, and I still need to check the paypal dialog too, but we're getting really close.

deBhal commented:
Ok! I've re-written most of it to lean on a httpData request, and suddenly it's all a whole lot simpler and seems to work well.

I'll go over it again in the morning when I'm fresh, but I think (hope) this is ready for review now.

deBhal commented:
Rebased

deBhal commented:
Rebased again, and made a subtle improvement to spacing in line with feedback in #28212:

> For the price buttons, I'd consider adding a bit more breathing room for readability. For example, I've added 2px top margin to the price, and 1px left margin to the superscript here. It's super subtle, but I think it helps everything look less cramped:

Before:

![price space before](https://user-images.githubusercontent.com/5952255/48178284-83118b80-e364-11e8-8d2d-efa9795bfd10.jpg)

After:
![price space after](https://user-images.githubusercontent.com/5952255/48178288-8573e580-e364-11e8-99a0-b0414a16872e.jpg)

southp commented on client/blocks/subscription-length-picker/index.jsx:
```diff
@@ -106,20 +138,26 @@ export class SubscriptionLengthPicker extends React.Component {
 }
 
-export function myFormatCurrency( price, code ) {
+export function myFormatCurrency( price, code, options = {} ) {
```
Totally soft opinion:

Would there be a better name for this function? I know it's been here for quite a while, but it would be great if it can have a more catchy name fitting its semantics, or even integrate with the underlying `formatCurrency` as a configurable behavior.

deBhal replied:
Yeah... It's dropping the cents if they're empty. `formatCurrency` uses the more general term 'fraction', so we could add an option named something like 'fractionOptional' or 'hideEmptyFraction'? `Intl.numberFormat` uses `minimumFractionDigits`, which isn't quite what we want (because we don't want to write `$12.3`, do we?).

How about this deal?: If you choose a good name for it, I'll move the functionality into a `formatCurrency()` option in a separate PR

southp replied:
Sounds good :)

southp commented on client/blocks/subscription-length-picker/index.jsx:
```diff
@@ -106,20 +138,26 @@ export class SubscriptionLengthPicker extends React.Component {
 
 	const hasCents = Math.abs( price % 1 ) >= EPSILON;
-	return formatCurrency( price, code, hasCents ? {} : { precision: 0 } );
+	return formatCurrency( price, code, hasCents ? options : { ...options, precision: 0 } );
```
It's a new behavior, so let's [extend its test suite](https://github.com/Automattic/wp-calypso/blob/master/client/blocks/subscription-length-picker/test/index.js#L81) to cover it.

southp commented on client/blocks/subscription-length-picker/index.jsx:
```diff
@@ -106,20 +138,26 @@ export class SubscriptionLengthPicker extends React.Component {
+	const paymentCountryCode = getPaymentCountryCode( state );
+	const paymentPostalCode = getPaymentPostalCode( state );
+	const taxRateRequest = requestTaxRate( paymentCountryCode, paymentPostalCode );
+	const taxRate = taxRateRequest.state === 'success' ? taxRateRequest.data : undefined;
```
Just a minor FYI, `taxRateRequest.data` will be `undefined` if it's not there, so this can be just `const taxRate = taxRateRequest.data` if I'm not mistaken.

southp commented on client/state/data-getters/index.js:
```diff
@@ -196,3 +200,49 @@ export const requestSitePost = ( siteId, postId, postType ) => {
+	}
+
+	const id = `tax-rate-${ countryCode }-${ postalCode }`;
+	const path = `/tax-rate/${ countryCode }/${ postalCode }` && '/sites/example.wordpress.com/hello'; // #tax-on-checout-placeholder
```
I personally would prefer you to stay consistent with the others here :)

southp commented on client/blocks/subscription-length-picker/index.jsx:
```diff
@@ -83,6 +111,10 @@ export class SubscriptionLengthPicker extends React.Component {
 									onCheck={ this.handleCheck }
+									shouldShowTax={ shouldShowTax }
+									taxDisplay={
+										shouldShowTax && this.formatTax( taxRate, priceFull, this.props.currencyCode )
```
Just like visibility flag of some common components, e.g. `<img>`, `<Button>`, `<div>`, I think visibility shouldn't affect passing data or not. e.g. This would look more conventional to me:
```
shouldShowTax={ shouldShowTax }
taxDisplay={ this.formatTax( ... ) }
```
However, this loses the benefit of not formatting the tax string if not showing, so the other way would be:
```
shouldShowTax={ ... }
taxRate={ ... }
priceFull={ ... }
currencyCode={ ... }
```
and let `<SubscriptionLengthOption>` to handle it underneath, including visibility and formatting the tax string or not.

Does that make sense to you?

deBhal replied:
Yeah... I wasn't thinking about hiding information from the child as much as not having the necessary data. Only components that should show the tax will have a `taxRate`. `formatTax()` short-circuits in this case, so let's go with your first suggestion.

I'm reluctant to push it down into the option itself, because it would need to get smarter to handle it, and it seems complicated enough already just deciding what to show and where to put it. I like the split between the "what" at the picker level and the "where" at the option level.

a8ci18n commented on client/blocks/subscription-length-picker/index.jsx:
```diff
@@ -42,11 +47,37 @@ export class SubscriptionLengthPicker extends React.Component {
+
+		// taxRate unknown
+		if ( ! isNumber( taxRate ) ) {
+			return translate( '+tax', {
```
Hi! I've found a possible matching string that has already been translated **54** times:
`translate( 'Tax' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

deBhal commented on client/state/selectors/get-payment-tax-rate.js:
```diff
@@ -0,0 +1,21 @@
+ * @return {?float} - The current combined tax rate, or null.
+ */
+export default function getPaymentTaxRate( state ) {
+	return get( state, 'ui.payment.taxRate', config.isEnabled( 'tax11' ) ? 0.11 : null );
```
It's a config flag I was using to insert a value for the tax formatting from the URL, but I think the `dummyTaxRate()` makes it superfluous, so now seems a good time to drop it.

deBhal commented:
Rebase

deBhal commented on client/blocks/subscription-length-picker/index.jsx:
```diff
@@ -42,11 +47,34 @@ export class SubscriptionLengthPicker extends React.Component {
+
+		// taxRate unknown
+		if ( ! isNumber( taxRate ) ) {
+			// translators: This string is displayed immediately next to a localized price with a currency symbol, and is indicating that there may be an additional charge on top of the displayed price.
```
> I'll have to check whether we already have a plan for this.

Yep - we're working towards using core's i18n tech, so this falls under that umbrella: https://github.com/Automattic/wp-calypso/pull/28504
